knitr/hmm-example/hmm-example.Rmd

```diff
@@ -0,0 +1,269 @@
+---
+title: "HMM Example"
+output: html_document
+---
+
+```{r setup, include=FALSE}
+library(tidyverse)
+library(ggplot2)
+library(cmdstanr)
+library(posterior)
+```
+
+## Introduction
+
+CmdStan 2.24 introduced a new interface for fitting Hidden Markov models (HMMs)
+in Stan. This document is intended to provide an example use of this interface.
+
+HMMs model a process where a system probabilistically switches between $K$
+states over a sequence of $N$ points in time. It is assumed that the exact
+state of the system is unknown and must be measured at each state.
+
+HMMs are characterized in terms of the transition matrix $\Gamma_{ij}$ (each
+element being the probability of transitioning from state $i$ to state $j$
+between measurements), the types of measurements made on the system (the
+system may emit continuous or discrete measurements), and the initial state
+of the system.
+
+Any realization of an HMM is a sequence of $N$ integers in the range $[1, K]$,
+however, because of the structure of the HMM, it is not necessary to sample
+the latent states to do inference on the transition probabilities, the
+parameters of the measurement model, or the estimates of the initial state.
+Estimates of the distribution of states at each measurement time can be
+computed separately.
+
+A more complete mathematical definition of the HMM model and function interface
+is given in the [Hidden Markov Models](https://mc-stan.org/docs/2_24/functions-reference/hidden-markov-models.html)
+section of the Function Reference Guide.
+
+There are three functions
+
+- `hmm_marginal` - The likelihood of an HMM with the latent discrete states
+integrated out
+- `hmm_latent_rng` - A function to generate random realizations from the posterior of an HMM
```

Comment:
Is it fair to say, "posterior random realizations" or something?

One thing that caught me off guard was I didn't know how to generate things from the prior for this -- like the log_omega argument is there!

Reply:
Also where to say hidden states and where to say latent states? Are they interchangeable or is there a difference somewhere?

Reply:
I guess realizations are not random by definition. Maybe the most precise term would be samples from the posterior. I'm blinking on why we needed the log omega argument there, but it definitely comes up in the algorithm. I would do prior simulations as you did in R.

I'd stick with hidden states to be consistent with the term hidden Markov models. They mean the same thing but latent could be more broadly applied.

Reply:
Since you're more familiar with the math, do you mind checking the `log_omega` thing?

Now that I think about it I don't understand why it's needed in `hmm_hidden_state_prob` either.

If I want to know the distribution of hidden states at time n of something with transition matrix A and initial state v, I just do v_n^T = v_0^T * A^n.

I'll go through and swap in hidden for latent!

Reply:
Whoops I'm assuming v there is a simplex, so it's not the initial state but the distribution of initial states.

Reply:
Ooooh, I finally realized what I was screwing up here. There is a difference between making draws from the integrated out states of the HMM and generated entirely new HMM trajectories. I added text to indicate that.